pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -386,52 +399,88 @@ type podVolume struct {
 // container is mounting the volume, the volume will be removed by desired state of world's populator and
 // cleaned up by the reconciler.
-func (rc *reconciler) reconstructStates(podsDir string) {
+func (rc *reconciler) syncStates(podsDir string) {
 	// Get volumes information by reading the pod's directory
 	podVolumes, err := getVolumesFromPodDir(podsDir)
 	if err != nil {
 		glog.Errorf("Cannot get volumes from disk %v", err)
 		return
 	}
+
+	volumesNeedUpdate := make(map[api.UniqueVolumeName]*reconstructedVolume)
 	for _, volume := range podVolumes {
-		volumeToMount, err := rc.reconstructVolume(volume)
+		reconstructedVolume, err := rc.reconstructVolume(volume)
 		if err != nil {
 			glog.Errorf("Could not construct volume information: %v", err)
 			continue
 		}
-
 		// Check if there is an pending operation for the given pod and volume.
 		// Need to check pending operation before checking the actual and desired
 		// states to avoid race condition during checking. For example, the following
 		// might happen if pending operation is checked after checking actual and desired states.
 		// 1. Checking the pod and it does not exist in either actual or desired state.
 		// 2. An operation for the given pod finishes and the actual state is updated.
 		// 3. Checking and there is no pending operation for the given pod.
-		if rc.operationExecutor.IsOperationPending(volumeToMount.VolumeName, volumeToMount.PodName) {
-			continue
-		}
-		desiredPods := rc.desiredStateOfWorld.GetPods()
-		actualPods := rc.actualStateOfWorld.GetPods()
-		if desiredPods[volume.podName] || actualPods[volume.podName] {
-			continue
+		// During state reconstruction period, no new volume operations could be issued. If the
+		// mounted path is not in either pending operation, or actual or desired states, this
+		// volume needs to be reconstructed back to the states.
+		pending := rc.operationExecutor.IsOperationPending(reconstructedVolume.volumeName, reconstructedVolume.podName)
+		dswExist := rc.desiredStateOfWorld.PodExistsInVolume(reconstructedVolume.podName, reconstructedVolume.volumeName)
+		aswExist, _, _ := rc.actualStateOfWorld.PodExistsInVolume(reconstructedVolume.podName, reconstructedVolume.volumeName)
+
+		// In case this is the first time to reconstruct state since kubelet starts, if the
+		// mounted path is for a persistant volume, it must have been mounted before kubelet
+		// restarts because no mount operations for persistent volumes could be started at this time (node
+		// status has not yet been updated before first syncStates finishes), In this case, the volume state
+		// should be put back to actual state now no matter desired state has it or not. Otherwise, the volume may not
+		// in the actual state when the mounted volume list is updated by node status updater and might cause master controller
+		// detaching while the volume is still mounted.
+		if !rc.StatesHasBeenSynced() {
+			if aswExist || !reconstructedVolume.pluginIsAttachable {
```

Comment:
The logic here is only if actual state already has the volume, it skip. Otherwise, no matter desired state has it or not, it will add the scanned volume to actual/desired state. So I am not quite sure about your question